components/raftstore/src/store/worker/refresh_config.rs

```diff
@@ -0,0 +1,214 @@
+use crate::store::fsm::apply::{ApplyFsm, ControlFsm};
+use crate::store::fsm::store::StoreFsm;
+use crate::store::fsm::PeerFsm;
+use batch_system::{BatchRouter, Fsm, FsmTypes, HandlerBuilder, Poller, PoolState, Priority};
+use file_system::{set_io_type, IOType};
+use std::fmt::{self, Display, Formatter};
+use std::sync::Arc;
+use std::thread;
+use tikv_util::worker::Runnable;
+use tikv_util::{debug, error, info, safe_panic, thd_name};
+
+pub struct PoolController<N: Fsm, C: Fsm, H: HandlerBuilder<N, C>> {
+    pub router: BatchRouter<N, C>,
+    pub state: PoolState<N, C, H>,
+}
+
+impl<N, C, H> PoolController<N, C, H>
+where
+    N: Fsm,
+    C: Fsm,
+    H: HandlerBuilder<N, C>,
+{
+    #[allow(dead_code)]
+    pub fn new(router: BatchRouter<N, C>, state: PoolState<N, C, H>) -> Self {
+        PoolController { router, state }
+    }
+}
+
+impl<N, C, H> PoolController<N, C, H>
+where
+    N: Fsm + std::marker::Send + 'static,
+    C: Fsm + std::marker::Send + 'static,
+    H: HandlerBuilder<N, C>,
+{
+    pub fn decrease_by(&mut self, size: usize) {
+        for _ in 0..size {
+            if let Err(e) = self.state.fsm_sender.send(FsmTypes::Empty) {
+                error!(
+                    "failed to decrese thread pool";
+                    "decrease to" => size,
+                    "err" => %e,
+                );
+                return;
+            }
+        }
+    }
+
+    pub fn increase_by(&mut self, size: usize) {
+        let name_prefix = self.state.name_prefix.clone();
+        let mut workers = self.state.workers.lock().unwrap();
+        self.state.id_base += size;
+        for i in 0..size {
+            let handler = self.state.handler_builder.build(Priority::Normal);
+            let mut poller = Poller {
+                router: self.router.clone(),
+                fsm_receiver: self.state.fsm_receiver.clone(),
+                handler,
+                max_batch_size: self.state.max_batch_size,
+                reschedule_duration: self.state.reschedule_duration,
+                before_pause_wait: self.state.before_pause_wait,
+                joinable_workers: Some(Arc::clone(&self.state.joinable_workers)),
+            };
+            let props = tikv_util::thread_group::current_properties();
+            let t = thread::Builder::new()
+                .name(thd_name!(format!(
+                    "{}-{}",
+                    name_prefix,
+                    i + self.state.id_base,
+                )))
+                .spawn(move || {
+                    tikv_util::thread_group::set_properties(props);
+                    set_io_type(IOType::ForegroundWrite);
+                    poller.poll();
+                })
+                .unwrap();
+            workers.push(t);
+        }
+    }
+
+    fn cleanup_poller_threads(&mut self) {
+        let mut joinable_workers = self.state.joinable_workers.lock().unwrap();
+        let mut workers = self.state.workers.lock().unwrap();
```

Comment:
Maybe we should add some comments to emphasize the order of acquiring these to mutexs

Reply:
Added, thanks!